File: examples/atb_models/tests/pythontest/operations_test/linear/test_linear.py
# Copyright (c) Huawei Technologies Co., Ltd. 2024. All rights reserved.
# MindIE is licensed under Mulan PSL v2.
# You can use this software according to the terms and conditions of the Mulan PSL v2.
# You may obtain a copy of Mulan PSL v2 at:
#          http://license.coscl.org.cn/MulanPSL2
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
# EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
# MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
# See the Mulan PSL v2 for more details.
import random
import unittest

import torch
import torch_npu

from operations_test import operation_test


class TestLinearOperation(operation_test.OperationTest):
    def setUp(self):
        self.op_type = "Linear"
        self.transpose_a = False
        self.transpose_b = False
        self.hasbias = False
        self.op_param = {
            "transposeA": self.transpose_a,
            "transposeB": self.transpose_b,
            "hasBias": self.hasbias
        }
        self.op_name = "LinearOperation"
        self.row = random.randint(1, 1000)
        self.col = random.randint(1, 1000)
        self.bias = 1e-5
        self.op_set = (self.op_type, self.op_param, self.op_name)

    def get_golden(self, in_tensor):
        golden_out_tensor = torch.add(torch.matmul(in_tensor['in0'], in_tensor['in1']), self.bias).npu()
        return [golden_out_tensor]

    def test_float16(self):
        in_tensor_0 = torch.rand(10, 3, dtype=torch.float16).npu()
        in_tensor_1 = torch.rand(3, 3, dtype=torch.float16).npu()
        out_tensor_0 = torch.zeros(10, 3, dtype=torch.float16).npu()

        inputs = {'in0': in_tensor_0, 'in1': in_tensor_1.t()}
        outputs = {'out0': out_tensor_0}

        self.run_compare(self.op_set, inputs, outputs)


if __name__ == '__main__':
    unittest.main()

File: examples/atb_models/tests/fuzztest/core/test_linear.cpp
/**
 * Copyright (c) Huawei Technologies Co., Ltd. 2024. All rights reserved.
 * MindIE is licensed under Mulan PSL v2.
 * You can use this software according to the terms and conditions of the Mulan PSL v2.
 * You may obtain a copy of Mulan PSL v2 at:
 *          http://license.coscl.org.cn/MulanPSL2
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
 * EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
 * MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
 * See the Mulan PSL v2 for more details.
 */

#include <gtest/gtest.h>
#include <atb/atb_infer.h>
#include "atb_speed/log.h"
#include "operations/fusion/linear/linear.h"
#include "../utils/fuzz_utils.h"
#include "secodeFuzz.h"

namespace atb_speed {
TEST(LinearDTFuzz, FusionLinear)
{
    std::srand(time(NULL));
    ATB_SPEED_LOG_DEBUG("begin====================");
    std::string fuzzName = "LinearDTFuzzFusionLinear";
    DT_FUZZ_START(0, 10000, const_cast<char*>(fuzzName.c_str()), 0) {
        uint32_t fuzzIndex = 0;
        int LinearQuantTypeEnumTable[] = {0, 1, 2, 3, 4, 5};
        int TransposeTypeEnumTable[] = {-1, 0, 1};
        atb_speed::common::FusionLinearParam param;
        param.quantType = static_cast<atb_speed::common::LinearQuantType>(
            *(unsigned int *) DT_SetGetNumberEnum(
                &g_Element[fuzzIndex++], 0, LinearQuantTypeEnumTable, 6)); // 6 is size of quant type
        param.isBF16 = FuzzUtil::GetRandomBool(fuzzIndex);;
        param.hasBias = FuzzUtil::GetRandomBool(fuzzIndex);;
        param.transposeType = static_cast<atb_speed::common::TransposeType>(
            *(int *) DT_SetGetNumberEnum(
                &g_Element[fuzzIndex++], -1, TransposeTypeEnumTable, 3)); // 3 is size of transpose type

        atb::Node linearNode;
        FusionLinear(param, &linearNode.operation);
    }
    DT_FUZZ_END()
    SUCCEED();
}
}